### PR TITLE
waybar: init at 0.4.0

### PR DESCRIPTION
**Pull request by FlorianFranzen**
###### Motivation for this change

Initial addition of [waybar](https://github.com/Alexays/Waybar).

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/misc/waybar/default.nix
Comment (alyssais) [inline pkgs/applications/misc/waybar/default.nix]: We don’t usually leave a newline here

File: pkgs/applications/misc/waybar/default.nix
Comment (alyssais) [inline pkgs/applications/misc/waybar/default.nix]: Nix code should be indented with two spaces.

File: pkgs/applications/misc/waybar/default.nix
Comment (alyssais) [inline pkgs/applications/misc/waybar/default.nix]: Perhaps put this next to the `buildInputs`. No need for it to be more than one line, either.

File: pkgs/applications/misc/waybar/default.nix
Comment (alyssais) [inline pkgs/applications/misc/waybar/default.nix]: You can take `lib` as an input and save yourself writing `stdenv.` over and over.

File: pkgs/applications/misc/waybar/default.nix
Comment (alyssais) [inline pkgs/applications/misc/waybar/default.nix]: There’s no need for this. You don’t need to provide default values for the package parameters.

File: pkgs/applications/misc/waybar/default.nix
Comment (alyssais) [inline pkgs/applications/misc/waybar/default.nix]: You can use `lib.optionalString` here.

File: pkgs/applications/misc/waybar/default.nix
Comment (alyssais) [inline pkgs/applications/misc/waybar/default.nix]: Can you explain what this does? Does waybar use the `sway` executable?

**Review by alyssais — CHANGES_REQUESTED** [review]
Lots of comments, but don’t worry. Mostly formatting.

File: pkgs/applications/misc/waybar/sysconfdir.patch
Comment (jtojnar) [inline pkgs/applications/misc/waybar/sysconfdir.patch]: This is actually correct and we would want [more applications](https://github.com/hughsie/fwupd/pull/824) to do this.

You should add `"-Dout=${placeholder "out"}"` and `"--sysconfdir=/etc"` to `mesonFlags`.

Comment (FlorianFranzen) [reply pkgs/applications/misc/waybar/sysconfdir.patch]: Thanks for the hint. Why do I need ```"--sysconfdir=/etc"```? Unless the ```sysconfdir``` is passed by meson to the applications, settings this should not make a difference.

Comment (jtojnar) [reply pkgs/applications/misc/waybar/sysconfdir.patch]: Yup, in this case sysconfdir will probably have no effect, as they use `g_get_system_data_dirs` to find the configuration.

**Comment by FlorianFranzen** [comment]
Update package file as requested, see rewritten commit.

File: pkgs/applications/misc/waybar/default.nix
Comment (FlorianFranzen) [reply pkgs/applications/misc/waybar/default.nix]: This is no longer needed, at least with sway-beta. The latest version of the bar does not work with sway < 1.0 anyway, so I dropped the the ```postInstall``` step completely.

File: pkgs/applications/misc/waybar/default.nix
Comment (alyssais) [inline pkgs/applications/misc/waybar/default.nix]: Commas go at the start of the following line, rather than the end of the line.

File: pkgs/applications/misc/waybar/default.nix
Comment (alyssais) [inline pkgs/applications/misc/waybar/default.nix]:
```suggestion
    version = "0.3.0";

```

File: pkgs/applications/misc/waybar/default.nix
Comment (alyssais) [inline pkgs/applications/misc/waybar/default.nix]:
```suggestion
      description = "Highly customizable Wayland bar for Sway and Wlroots based compositors";
```

From the [manual](https://nixos.org/nixpkgs/manual/#sec-standard-meta-attributes):

>  Don’t include a period at the end. Don’t include newline characters. Capitalise the first character. For brevity, don’t repeat the name of package — just describe what it does.

**Review by alyssais — CHANGES_REQUESTED** [review]
Couple more style notes.

File: pkgs/applications/misc/waybar/default.nix
Comment (alyssais) [inline pkgs/applications/misc/waybar/default.nix]: What's the closure size impact of just defaulting all of these to true?

Comment (FlorianFranzen) [reply pkgs/applications/misc/waybar/default.nix]: According to ```nix path-info -S ...```, this would change the size from 287,863,928 to 294,567,016.

Comment (alyssais) [reply pkgs/applications/misc/waybar/default.nix]: That's not a big difference. Do you think we should just enable all of them?

File: pkgs/top-level/all-packages.nix
Comment (alyssais) [inline pkgs/top-level/all-packages.nix]:
```suggestion

```

**Review by alyssais — CHANGES_REQUESTED** [review]

**Comment by FlorianFranzen** [comment]
@alyssais Now trailing whitespace free.

**Comment by FlorianFranzen** [comment]
Now also includes a NixOS module to run waybar as systemd service. 

@primeos I am interested in implement sway's recommed [systemd integration](https://github.com/swaywm/sway/wiki/Systemd-integration) in NixOS. Any feedback to this module would be welcome, especially on how to pass the ```SWAYSOCK``` variable to waybay and others (without a call to ```systemctl --user import-environment ...```).

**Comment by FlorianFranzen** [comment]
@alyssais Is there anything else missing before we can get this merged?

**Comment by primeos** [comment]
> @primeos I am interested in implement sway's recommed [systemd integration](https://github.com/swaywm/sway/wiki/Systemd-integration) in NixOS. Any feedback to this module would be welcome, especially on how to pass the `SWAYSOCK` variable to waybay and others (without a call to `systemctl --user import-environment ...`).

Sorry for my late response, I'm not a huge fan of systemd user units (at least ATM) and therefore also not that familiar with them.

There should be multiple ways to pass `SWAYSOCK` but I'm not sure which one is the best. It would be nice if `SWAYSOCK` could be inherited by having a dependency on `sway.service` but I assume this isn't the case. It should be possible to use `systemctl --user import-environment SWAYSOCK` from withing `sway.service` but that solution could get a bit complicated (doesn't seem like the worst idea though). Another approach would be to hard-code `SWAYSOCK` but this will become problematic if users need multiple Sway sessions (though that most likely won't work well with systemd anyway) and might cause other problems. Using systemd environment file generators (`man systemd.environment-generator`) should also be possible for a more dynamic approach.

But in any case it might be best to first check how other units are solving this problem.

**Comment by alyssais** [comment]
@FlorianFranzen I'm happy to merge this once I see an ack of @primeos's comment from you. Let me know if you plan to change it further – I was about to merge it before and then a whole new module appeared!

**Comment by FlorianFranzen** [comment]
In do not want to make ```waybar``` sway-only as it is not by design. That is why I made it depend on the ```graphical-session.target``` instead, which should be the correct and general way (see e.g. #22671).

@primeos Thank you for your feedback. As ```SWAYSOCK``` is needed inside a sway session for the IPC message client to function anyway, any systemd solution will have to take care of this as well. If I find some time to clean up my solution, I will make sure to open a pull request and let you know.

Hardcoding ```SWAYSOCK``` is not sensible solution as the name always contains the process id of sway (i.e. ```/run/user/1000/sway-ipc.1000.1547.sock```).

@alyssais I have no plans to change this further, thank you for your help to get this merged.

**Comment by FlorianFranzen** [comment]
Updated waybar to latest release.

**Comment by alyssais** [comment]
Sorry for how slow I've been on this.

I just took a look, and it needs to be updated, because with sway 1.0,
there's no longer a sway-beta package.

**Comment by FlorianFranzen** [comment]
@alyssais: No problem, that was bound to happen. I rebased the pull request against the latest unstable and replaced all reference to ```sway-beta``` accordingly.

**Comment by alyssais** [comment]
Thank you for your persistence!

**Pull request merged**
